os/se/ameba/rtl_security_api_wrapper.c

```diff
@@ -0,0 +1,484 @@
+		}
+	} else {
+		ret = RTL_HAL_FAIL;
+	}
+	device_mutex_unlock(RT_DEV_LOCK_CRYPTO);
+
+	return ret;
+}
+
+uint32_t rtl_cryptoAES_cbc_wrapper(uint8_t* key, uint32_t keylen, unsigned char* message, uint32_t msglen, unsigned char* pResult, uint8_t mode)
+{
+	uint32_t ret = RTL_HAL_SUCCESS;
+	unsigned char iv[16];
+
+	memcpy(iv, aes_iv, sizeof(aes_iv));
+	device_mutex_lock(RT_DEV_LOCK_CRYPTO);
+	ret = rtl_crypto_aes_cbc_init(key, keylen);
+
+	if (ret == RTL_HAL_SUCCESS) {
+		if (mode == RTL_ENCRYPT) {
+			/* Encryption operation */
+			ret = rtl_crypto_aes_cbc_encrypt(message, msglen, iv, sizeof(iv), pResult);
+		} else if (mode == RTL_DECRYPT) {
+			/* Decryption operation */
+			ret = rtl_crypto_aes_ecb_decrypt(message, msglen, iv, sizeof(iv), pResult);
+		} else {
+			ret = RTL_HAL_NOT_SUPPORTED;
+		}
+	} else {
+		ret = RTL_HAL_FAIL;
+	}
+	device_mutex_unlock(RT_DEV_LOCK_CRYPTO);
+
+	return ret;
+}
+
+uint32_t rtl_cryptoAES_ctr_wrapper(uint8_t* key, uint32_t keylen, unsigned char* message, uint32_t msglen, unsigned char* pResult, uint8_t mode)
+{
+	uint32_t ret = RTL_HAL_SUCCESS;
+	unsigned char iv[16];
+
+	memcpy(iv, aes_iv, sizeof(aes_iv));
+	device_mutex_lock(RT_DEV_LOCK_CRYPTO);
+	ret = rtl_crypto_aes_ctr_init(key, keylen);
+
+	if (ret == RTL_HAL_SUCCESS) {
+		if (mode == RTL_ENCRYPT) {
+			/* Encryption operation */
+			ret = rtl_crypto_aes_ctr_encrypt(message, msglen, iv, sizeof(iv), pResult);
+		} else if (mode == RTL_DECRYPT) {
+			/* Decryption operation */
+			ret = rtl_crypto_aes_ctr_decrypt(message, msglen, iv, sizeof(iv), pResult);
+		} else {
+			ret = RTL_HAL_NOT_SUPPORTED;
+		}
+	} else {
+		ret = RTL_HAL_FAIL;
+	}
+	device_mutex_unlock(RT_DEV_LOCK_CRYPTO);
+
+	return ret;
+}
+
+/**
+ * Factory Key and Cert
+ */
+uint32_t rtl_read_factory_cert_wrapper(hal_data *data)
+{
+	uint32_t ret;
+	uint32_t data_len;
+	uint8_t buf_length[4];
+	uint8_t *buf_flash = NULL;
+	flash_t flash;
+
+	device_mutex_lock(RT_DEV_LOCK_FLASH);
+	flash_stream_read(&flash, FACTORY_CERT_ADDR, 4, buf_length);
+	device_mutex_unlock(RT_DEV_LOCK_FLASH);
+
+	data_len = atoi(buf_length);	/* Cert Length */
+
+	buf_flash = (uint8_t *) malloc(data_len + 4);	/* Allocate buff */
+
+	if (buf_flash) {
+		memset(buf_flash, 0, data_len + 4);
+
+		device_mutex_lock(RT_DEV_LOCK_FLASH);
+		flash_stream_read(&flash, FACTORY_CERT_ADDR, data_len + 4, buf_flash);
+		device_mutex_unlock(RT_DEV_LOCK_FLASH);
+
+		memcpy(data->data, buf_flash + 4, data_len);	/* Copy Cert into output buff */
+		data->data_len = data_len;
+
+		ret = RTL_HAL_SUCCESS;
+	} else {
+		ret = RTL_HAL_FAIL;
+	}
+
+	if (buf_flash != NULL) {
+		free(buf_flash);
+	}
+
+	return ret;
+}
+
+uint32_t rtl_read_factory_key_wrapper(hal_data *data)
+{
+	uint32_t ret;
+	uint32_t data_len;
+	uint8_t buf_length[4];
+	uint8_t *buf_plain = NULL;
+	uint8_t *buf_crypt = NULL;
+	uint8_t *buf_flash = NULL;
+	flash_t flash;
+	unsigned char aes_key[32] __attribute__((aligned(4)));
+
+	device_mutex_lock(RT_DEV_LOCK_FLASH);
```

Comment:
where is definition of device_mutex_lock. I couldn't find it

Reply:
os/board/rtl8721csm/src/component/os/os_dep/include/device_lock.h and os/board/rtl8721csm/src/component/os/os_dep/device_lock.c